### PR TITLE
Solves #404 Refactor dao

### PR DESCRIPTION
**Pull request by npathai**
List of changes:
 - Two implementations of `CustomerDao` interface. `DbCustomerDao` and `InMemoryCustomerDao`.
 - `CustomerDaoImpl` is renamed to `InMemoryCustomerDao`.
 - Use of Hierarchical Context Runner for making tests more readable.
 - Updated class diagram.
 - Schema is created by App before using `DbCustomerDao`.

### CODE REVIEW

**Comment by npathai** [comment]
@iluwatar Please go through the changes. I have changed test cases a bit to use my style of unit testing. The method names are verbose and communicate the behavior under test.

File: dao/pom.xml [hunk 1]
```diff
@@ -44,6 +44,19 @@
       <groupId>log4j</groupId>
       <artifactId>log4j</artifactId>
     </dependency>
+    <dependency>
+        <groupId>com.h2database</groupId>
+        <artifactId>h2</artifactId>
+        <scope>compile</scope>
```
Comment (iluwatar) [inline dao/pom.xml]: `compile` is the default scope so no need to define it

**Comment by iluwatar** [comment]
Review remarks:
- The SQL for creating/dropping schema is duplicated multiple times here and there. Would it be possible to put this in one place e.g. SQL-file and execute from there?
- Handling of `null` values. It's a matter of style and opinion but we could utilize `Optional` instead.

**Comment by iluwatar** [comment]
@npathai you've got my review comments.

**Comment by npathai** [comment]
@iluwatar 
 - Regarding SQL I will try to address it.
 - Regarding `Optional`, I gave it some thought while designing the interface according to java 8. But left it as is because Optional is useful when `null` can have two meanings. Such as `Map.get()`, where it can mean that key was not mapped to any value or key was mapped to a null value. There were a few links as well discussing it. I will attach it. Optional is not a replacement for null.

**Comment by npathai** [comment]
@iluwatar Well after a lot of opinion reading I decided that for the sake of keeping interface fuctional ready, I will change the return type to be Optional.

But in reality I dont hate nulls. If documented correctly they are not a problem. Also creating an optional wrapper creates unnecessary GC pressure if the operation is heavily used.

**Comment by npathai** [comment]
@iluwatar Done with the changes.

**Comment by iluwatar** [comment]
@npathai these are great improvements :+1:

**Pull request merged**